Adding the content for "Some issues" (#4)
citynetwork/ois2023: markdown/issues.md

And to teardown the associated volume, it tries to remove the mount directory. It tries to run os.remove method on the volume path, and if something goes wrong while this execution volume path is left on the node. However mount path is deleted and when kubelet retries to unmount the volume, it will fail unmount operation as it will not be able to find the mount path.

This problem of pods getting stuck in the "Terminating" state due to the inability to clean volume subPath mounts is a known issue in Kubernetes. Several users have reported similar issues and there is no definite cause what arises the issue.

The workaround suggested in the bug reports was to manually delete the orphaned pod directory and restart the kubelet service, which works for us also. But we have a CI driven deployment and it's very complicated to manually shell into nodes and deleting stale directories when the pipeline is stuck on terminating (orphaned) pods.

Ideally, kubelet should intelligently deal with the orphaned pods, cleaning a stale directory manually should not be required.

> As per the bug reports the orphaned pods issue is consistent until kuberntes version 1.24 atleast and maybe are fixed in the latest realeases, which brings us to our next issue.

Reviewer: ```suggestion
As per the bug reports, the orphaned pods issue is present in Kubernetes versions up to at least 1.24. It is *possibly* fixed in later releases, which brings us to our next issue.
```

## Kubernetes versions

<!-- Note -->

> In our cloud infrastructure, we currently have OpenStack Xena deployed. This version of OpenStack only supports Kubernetes upto version 1.21.x and Fedora CoreOS 31.

Reviewer: ```suggestion
In our cloud infrastructure, we currently have OpenStack Xena deployed. This version of OpenStack only supports Kubernetes up to version 1.21.x, and "officially" only supports Fedora CoreOS 31 (although we at least got it to run on Fedora CoreOS 34).
```

> As we are experiencing issues with orphaned pods and the solutions available in newer Kubernetes versions can't be deployed and we are using Fedora CoreOS 31 which has reached its EOL,  which means it is no longer supported and may not receive critical updates and security patches. 

Reviewer: We are not using FCOS 31, we are using 34.

> If we upgrade from OpenStack Xena to Openstack Antelope also, we get support for Kubernetes version v1.24 which will not fix the Orphaned pod issue and will also be end of life in few days (2023-07-28).

Reviewer: You should check with a Magnum developer here if 1.24 is *really* the latest version that can be made to work with Antelope.

Author: Okay, I will ask this question on IRC and see if I get an answer.

* The second option, NodePort, is also unsuitable for our needs since our Caddy service operates on ports 80 and 443, while the NodePort range is from 30000 to 32767.

* The third option is a loadbalancer, which we are trying to eliminate.
* Lastly, there is a way not widely used among the kubernetes community because many are using cloud provider's Load balancers.
Using ClusterIP with external IP

Customers can expose their Kubernetes cluster to external traffic using ClusterIP and an external IP address. The Kubernetes service type is ClusterIP, but an external IP address can be specified using the externalIP property. This solution requires that the customer maps the external IP address of a service to their private IP addresses in the cluster. 

Unfortunately, we encountered issues in utilizing a Floating IP address as an external IP address. It was due to the fact that our Floating IP address was mapped to a private address within a different subnet than the one used by the ClusterIP.

> So, we were not able to figure out any way to expose our services without Loadbalancer.

Reviewer: Well, no we weren't, but we _were_ able to come up with an approach where rather than doing anything with the Kubernetes APIs, we would just operate on a single DNS record that we could flip, without using any API calls, to the other site. That allows us to have at least a rudimentary service on the backup site, until load balancer functionality in the failed primary site becomes available again.

## Orphaned Pods

<!-- Note -->

What are Orphaned pods in a Kubernetes cluster?

Orphaned pods can occur in Kubernetes when their owner objects, such as a deployment, replica set, or replication controller, is deleted or modified.
By default, Kubernetes deletes these dependent objects. Therefore, the responsibility of cleaning up these objects lies with the kubelet.

What is kubelet?

The kubelet is the primary "node agent" that runs on each node.
Kubelet is responsible for applying, creating, updating, and destroying containers on a Kubernetes node. It performs garbage collection on unused images and unused containers on the nodes.

When kubelet fails to delete these dependent objects, orphaned pods remain dangling in the environment and interfering with the deployment process.

We are facing this issue while performing a rolling update in our deployment.

What happens in a rolling deployment?
During a rolling deployment, the controller deletes and recreates pods with the updated configuration, one by one, without causing downtime to the cluster.

> However, our rolling deployment gets stuck when the pod running the old version remains in the Terminating state, and the new version that needs to be recreated gets stuck in the ContainerCreating state.

Reviewer: You must add here that this is only a problem because some of the affected Pods mount volumes that are `RWO`, rather than `RWX`. If we could use `RWX` volumes, this wouldn't be a problem — and it *isn't* more than a cosmetic problem for Pods not using any volumes at all (of which we also have several)
